src/modules/commander/commander.cpp

```diff
@@ -3707,6 +3712,216 @@ set_main_state_rc(struct vehicle_status_s *status_local)
+		last_gpos_fail_time_us = hrt_absolute_time();
+	}
+
+	// check global velocity accuracy with hysteresis in both test level and time
+	bool vel_status_changed = false;
+	if (status_flags.condition_global_velocity_valid && global_position->evh > evh_threshold * 2.5f) {
+		global_vel_inaccurate = true;
+		vel_status_changed = true;
+		last_gvel_fail_time_us = hrt_absolute_time();
+	} else if (!status_flags.condition_global_velocity_valid) {
+		bool check_level_pass = global_position->evh < evh_threshold;
+		if (!check_level_pass || hold_fail_state) {
+			gvel_probation_time_us += (hrt_absolute_time() - last_gvel_fail_time_us) * POSVEL_VALID_PROBATION_FACTOR;
+			last_gvel_fail_time_us = hrt_absolute_time();
+		} else if (hrt_absolute_time() - last_gvel_fail_time_us > gvel_probation_time_us) {
+			global_vel_inaccurate = false;
+			vel_status_changed = true;
+			last_gvel_fail_time_us = 0;
+		}
+	} else {
+		gvel_probation_time_us -= (hrt_absolute_time() - last_gvel_fail_time_us);
+		last_gvel_fail_time_us = hrt_absolute_time();
+	}
+
+	bool global_data_stale = (hrt_absolute_time() - global_position->timestamp > POSITION_TIMEOUT);
+
+	// Set global velocity validity
+	if (vel_status_changed) {
+		if (status_flags.condition_global_velocity_valid
+			   && (global_data_stale || global_vel_inaccurate)) {
+			status_flags.condition_global_position_valid = false;
+			*changed = true;
+		} else if (!status_flags.condition_global_velocity_valid
+			   && !global_data_stale
+			   && !global_vel_inaccurate) {
+			status_flags.condition_global_velocity_valid = true;
+			*changed = true;
+		}
+	}
+
+	// Set global position validity
+	if (pos_status_changed) {
+		if (status_flags.condition_global_position_valid
+			   && (global_data_stale || global_pos_inaccurate)) {
+			status_flags.condition_global_position_valid = false;
+			set_tune_override(TONE_GPS_WARNING_TUNE);
+			*changed = true;
+		} else if (!status_flags.condition_global_position_valid
+			   && !global_data_stale
+			   && !global_pos_inaccurate) {
+			status_flags.condition_global_position_valid = true;
+			*changed = true;
+		}
+	}
+
+	// constrain probation times
+	if (land_detector.landed) {
+		gpos_probation_time_us = POSVEL_PROBATION_MIN;
+		gvel_probation_time_us = POSVEL_PROBATION_MIN;
+	} else {
+		if (gpos_probation_time_us < POSVEL_PROBATION_MIN) {
+			gpos_probation_time_us = POSVEL_PROBATION_MIN;
+		} else if  (gpos_probation_time_us > POSVEL_PROBATION_MAX) {
+			gpos_probation_time_us = POSVEL_PROBATION_MAX;
+		}
+		if (gvel_probation_time_us < POSVEL_PROBATION_MIN) {
+			gvel_probation_time_us = POSVEL_PROBATION_MIN;
+		} else if  (gvel_probation_time_us > POSVEL_PROBATION_MAX) {
+			gvel_probation_time_us = POSVEL_PROBATION_MAX;
+		}
+	}
+}
+
+void
+check_local_posvel_validity(struct vehicle_local_position_s *local_position, bool *changed)
+{
+	bool local_pos_inaccurate = false;
+	bool local_vel_inaccurate = false;
+	bool hold_fail_state = false;
```

Comment:
same comments apply here too as above.
In addition this is pretty much duplicated and it would be nice if we could merge the two. The same holds for the position vs velocity within the method. For example with one generic method and providing it with the necessary pointers. And maybe create a struct if the number of arguments explodes?